x-pack/plugins/actions/server/actions_client.ts

```diff
@@ -111,20 +118,45 @@ export class ActionsClient {
   public async create({
     action: { actionTypeId, name, config, secrets },
   }: CreateOptions): Promise<ActionResult> {
-    await this.authorization.ensureAuthorized('create', actionTypeId);
+    const id = SavedObjectsUtils.generateId();
+
+    try {
+      await this.authorization.ensureAuthorized('create', actionTypeId);
+    } catch (error) {
+      this.auditLogger?.log(
+        connectorEvent({
+          action: ConnectorAction.CREATE,
+          savedObject: { type: 'action', id },
+          error,
+        })
+      );
+      throw error;
+    }
```

Comment:
As @gmmorris mentioned, would some of these fit better within the `this.authorization.ensureAuthorized` call?

Reply:
We could potentially move authorisation errors into the `ensureAuthorized` method. However, that does not work for the authorisation success scenario since we need to log which saved objects were accessed and we don't know that until we received the response from Elasticsearch. Have a look at the `find` operation were this is particular obvious.

I prefer to have the success/error audit events in the same place rather than dispersed in different services.

Reply:
I see what you mean, half of the logic would live in each location. I see this pattern used elsewhere like the [security plugin SO wrapper](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts) so I'm 👍 leaving it as in then.